src/shims/windows/fs.rs

```diff
@@ -0,0 +1,334 @@
+            desired_access &= !generic_read;
+            options.read(true);
+        }
+        if desired_write {
+            desired_access &= !generic_write;
+            options.write(true);
+        }
+
+        if desired_access != 0 {
+            throw_unsup_format!(
+                "CreateFileW: Unsupported bits set for access mode: {desired_access:#x}"
+            );
+        }
+
+        // This is racy, but there doesn't appear to be an std API that both succeeds if a file
+        // exists but tells us it isn't new. Either we accept racing one way or another, or we
+        // use an iffy heuristic like file creation time. This implementation prefers to fail in the
+        // direction of erroring more often.
+        let exists = file_name.exists();
+
+        if creation_disposition == create_always {
+            // Per the documentation:
+            // If the specified file exists and is writable, the function truncates the file, the
+            // function succeeds, and last-error code is set to ERROR_ALREADY_EXISTS.
+            // If the specified file does not exist and is a valid path, a new file is created, the
+            // function succeeds, and the last-error code is set to zero.
+            // https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew
+            if exists {
+                this.set_last_error(IoError::WindowsError("ERROR_ALREADY_EXISTS"))?;
+            } else {
+                this.set_last_error(IoError::Raw(Scalar::from_u32(0)))?;
+            }
+            options.create(true);
+            options.truncate(true);
+        } else if creation_disposition == create_new {
+            options.create_new(true);
+            // Per `create_new` documentation:
+            // If the specified file does not exist and is a valid path to a writable location, the
+            // function creates a file and the last-error code is set to zero.
+            // https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.create_new
+            if !desired_write {
+                options.append(true);
+            }
+        } else if creation_disposition == open_always {
+            // Per the documentation:
+            // If the specified file exists, the function succeeds and the last-error code is set
+            // to ERROR_ALREADY_EXISTS.
+            // If the specified file does not exist and is a valid path to a writable location, the
+            // function creates a file and the last-error code is set to zero.
+            // https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew
+            if exists {
+                this.set_last_error(IoError::WindowsError("ERROR_ALREADY_EXISTS"))?;
+            } else {
+                this.set_last_error(IoError::Raw(Scalar::from_u32(0)))?;
+            }
+            options.create(true);
+        } else if creation_disposition == open_existing {
+            // Nothing
```

Comment:
Are you sure this does not reset the error code to 0?

Maybe we should just set it to 0 at the top, to make sure we don't forget?